libraries/chain/db_market.cpp

```diff
@@ -911,22 +911,58 @@ bool database::fill_call_order( const call_order_object& order, const asset& pay
 
+/***
+ * @brief fullfill a settle order in the specified amounts
+ *
+ * @details Called from database::match(), this coordinates exchange of debt asset X held in the
+ *    settle order for collateral asset Y held in a call order, and routes fees.  Note that we
+ *    don't touch the call order directly, as match() handles this via a separate call to
+ *    fill_call_order().  We are told exactly how much X and Y to exchange, based on details of
+ *    order matching determined higher up the call chain. Thus it is possible that the settle
+ *    order is not completely satisfied at the conclusion of this function.
+ *
+ * @param settle the force_settlement object
+ * @param pays the quantity of market-issued debt asset X which the settler will yield in this
+ *    round (may be less than the full amount indicated in settle object)
+ * @param receives the quantity of collateral asset Y which the settler will receive in
+ *    exchange for X
+ * @param fill_price the price at which the settle order will execute (not used - passed through
+ *    to virtual operation)
+ * @param is_maker TRUE if the settle order is the maker, FALSE if it is the taker (passed
+ *    through to virtual operation)
+ * @returns TRUE if the settle order was completely filled, FALSE if only partially filled
+ */
 bool database::fill_settle_order( const force_settlement_object& settle, const asset& pays, const asset& receives,
                                   const price& fill_price, const bool is_maker )
 { try {
    bool filled = false;
 
    const account_object* settle_owner_ptr = nullptr;
-   // The owner of the settle order pays market fees to the issuer of the collateral asset after HF core-1780
+   // The owner of the settle order pays market fees to the issuer of the collateral asset.
+   // After HF core-1780, these fees are shared to the referral program, which is flagged to
+   // pay_market_fees by setting settle_owner_ptr non-null.
    //
    // TODO Check whether the HF check can be removed after the HF.
    //      Note: even if logically it can be removed, perhaps the removal will lead to a small performance
    //            loss. Needs testing.
    if( head_block_time() >= HARDFORK_CORE_1780_TIME )
       settle_owner_ptr = &settle.owner(*this);
+   // Compute and pay the market fees:
+   asset market_fees = pay_market_fees( settle_owner_ptr, get(receives.asset_id), receives );
+
+   // Issuer of the settled smartcoin asset lays claim to a force-settlement fee (BSIP87), but
+   // note that fee is denominated in collateral asset, not the debt asset.  Asset object of
+   // debt asset is passed to the pay function so it knows where to put the fee. Note that
+   // amount of collateral asset upon which fee is assessed is reduced by market_fees already
+   // paid to prevent the total fee exceeding total collateral.
+   asset force_settle_fees = pay_force_settle_fees( get(pays.asset_id), receives - market_fees );
```

Comment:
> IMO the force-settlement fee should be paid before the market fees, since the operation's main business is the exchanging of the debt and the collateral, and the force-settlement fee is a part of the business logic. In comparison, the market fee is much more passive here.

I agree with my reading of your comment: I do not consider force-settlement to be a market exchange because it is, in my opinion, a transaction between the settler and the blockchain.

But my opinion seems to be contradicted by the fact the _existing_ code does charge a market fee for force-settlement, which implies that the code considers the transaction to be a market exchange.  For whatever it's worth.

Reply:
> So it really hinges on exactly what the scope of the market fee is and what activities are intended to be subjected to it.
> 
> Tagging @MichelSantos in here for some business analytic expertise. This is an interesting question.

I think that the proposed implementation is consistent with two points in the specifcation

1. The sequencing of the multiplication in the formula implies the sequencing in the proposed implementation

2. The [prose](https://github.com/bitshares/bsips/blob/master/bsip-0081.md) can be interpreted to mean that the "delta" force settlement fee is the second calculation (emphasis added)

> When a force settlement is executed, the buyer sells smartcoin with quantity `X` and get collateral in quantity `X*(1-FSO)*(1-FSFP)/feed price price*`, the settled debt position owner sells collateral in quantity `X*(1-FSO)/feed price*` and get smartcoin in quantity `X`, the **delta** and received collateral in quantity `X*FSFP*(1-FSO)/feed price*` will be paid to the owner of the smartcoin as force settlement fee.

I think that the proposed implementation is a reasonable reading of the slightly vague specification.

Reply:
You convinced me. I agree that the current implementation makes sense.